### PR TITLE
Issue #1814: Enhance state mismatch error handling and messaging

### LINKED ISSUE #1814: Login still resulting in "401" error for some users
### Current Behavior

Some users are still seeing a "401" error when logging in

1. "401" error is displayed when user logs in
2. By clicking sign in again, user is able to successfully log in without error

[Slack discussion](https://cisa-corp.slack.com/archives/C05BGB4L5NF/p1708556270444239)

Logs recorded:
image

### Expected Behavior

User logs in successfully without encountering error.

### Steps to Reproduce
Users who see this just do the following
1. click "sign in"
2. "401" error is displayed
3. By clicking sign in again, user is able to successfully log in without error

Katherine is the one on the team who sees this with frequency so it would be worth seeing the logs around when her email logs in.
### Environment

Environment: Stable
Browser: Chrome
Login method: username/password combo

### Additional Context
We fixed this issue in relation to it being related to states and we appear to have another bug that we missed with OIDC that is presenting similar behavior to the last one
related to: #1505

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #1814 

## Changes

- Attempt a single redirect after any state mismatch error.

## Context for reviewers

Didn't think about unit tests yet. Or how to test this in general.

I do have elaborate logging for state mismatch errors, which we can remove after a few weeks if this thing proves hard to test and debug otherwise.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/djangooidc/oidc.py [hunk 1]
```diff
@@ -182,7 +182,11 @@ def callback(self, unparsed_response, session):
 
         if authn_response["state"] != session.get("state", None):
             # this most likely means the user's Django session vanished
-            logger.error("Received state not the same as expected for %s" % state)
+            logger.error(
+                f"Received state not the same as expected for {state}"
+                f"authn_response['state'] = {authn_response['state']}"
+                f"session.get('state', None) = {session.get('state', None)}"
+            )
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/djangooidc/oidc.py]: Yep. Converting this into an actual PR, stay tuned.

File: src/djangooidc/views.py [hunk 2]
```diff
@@ -108,16 +108,24 @@ def login_callback(request):
...
-        # beginning of login process without raising an error to the user.
-        logger.warning(f"No State Defined: {nsd_err}")
-        return redirect(request.session.get("next", "/"))
+    except o_e.StateMismatch as nsd_err:
+        # Check if the redirect has already been attempted
+        if not request.session.get("redirect_attempted", False):
+            # Set the flag to indicate that the redirect has been attempted
+            request.session["redirect_attempted"] = True
+
+            # In the event of a sstate mismatch between OP and session, redirect the user to the
```
Comment (abroddrick) [inline src/djangooidc/views.py:122]:
```suggestion
            # In the event of a state mismatch between OP and session, redirect the user to the
```

File: src/djangooidc/views.py [hunk 3]
```diff
@@ -108,16 +108,24 @@ def login_callback(request):
...
+        if not request.session.get("redirect_attempted", False):
+            # Set the flag to indicate that the redirect has been attempted
+            request.session["redirect_attempted"] = True
+
+            # In the event of a state mismatch between OP and session, redirect the user to the
+            # beginning of login process without raising an error to the user. Attempt once.
+            logger.warning(f"No State Defined: {nsd_err}")
+            return redirect(request.session.get("next", "/"))
+        else:
+            # Clear the redirect_attempted flag?
```
Comment (abroddrick) [inline src/djangooidc/views.py:127]: I get the question here, I think I flipflopped a few times just staring here. I vote safer to just ensure no matter route the flag is remove. So yes, I suggest just removing the flag if it still exists for any reason (granted it really should _not_ ever happen)

Comment (rachidatecs) [reply src/djangooidc/views.py:127]: Yep, sounds good to me

File: src/djangooidc/tests/test_views.py [hunk 4]
```diff
@@ -129,21 +129,33 @@ def test_openid_initializes_client_and_calls_create_authn_request(self, mock_cli
...
-            # ASSERTIONS
-            # assert that the user is redirected to the start of the login process
+            # ASSERT
             self.assertEqual(response.status_code, 302)
             self.assertEqual(response.url, "/")
+            # Check that the redirect_attempted flag is set in the session
+            self.assertTrue(self.client.session.get("redirect_attempted", False))
+
+    def test_login_callback_with_no_session_state_attempt_again_only_once(self, mock_client):
+        """We only attempt to relogin once. After that, it's the error page for you."""
```
Comment (abroddrick) [inline src/djangooidc/tests/test_views.py:147]: Lol i'm going to pretend I didn't read this so I don't have to suggest changing it, lol-ed at "it's the error page for you"

File: src/djangooidc/tests/test_views.py [hunk 5]
```diff
@@ -129,21 +129,33 @@ def test_openid_initializes_client_and_calls_create_authn_request(self, mock_cli
...
+        with less_console_noise():
+            # MOCK
```
Comment (abroddrick) [inline src/djangooidc/tests/test_views.py:149]:
```suggestion
            # MOCK- the session and error
```

**non-blocking suggestion**

File: src/djangooidc/tests/test_views.py [hunk 6]
```diff
@@ -129,21 +129,33 @@ def test_openid_initializes_client_and_calls_create_authn_request(self, mock_cli
...
+            mock_client.get_default_acr_value.side_effect = self.create_acr
+            mock_client.callback.side_effect = StateMismatch()
+            session = self.client.session
+            session["redirect_attempted"] = True
+            session.save()
+            # TEST
```
Comment (abroddrick) [inline src/djangooidc/tests/test_views.py:155]:
```suggestion
            # TEST- receiving a response from login.gov
```

**non-blocking suggestion**

File: src/djangooidc/tests/test_views.py [hunk 7]
```diff
@@ -129,21 +129,33 @@ def test_openid_initializes_client_and_calls_create_authn_request(self, mock_cli
                     self.assertContains(response, "Hi")
 
     def test_login_callback_with_no_session_state(self, mock_client):
-        """If the local session is None (ie the server restarted while user was logged out),
+        """If the local session does not match the OP session,
         we do not throw an exception. Rather, we attempt to login again."""
         with less_console_noise():
             # MOCK
```
Comment (abroddrick) [inline src/djangooidc/tests/test_views.py:135]: **non blocking suggestion**: I like the old comment better
```suggestion
            # MOCK the callback to raise StateMismatch error when called
```

File: src/djangooidc/views.py [hunk 8]
```diff
@@ -108,16 +108,25 @@ def login_callback(request):
...
+            request.session["redirect_attempted"] = True
+
+            # In the event of a state mismatch between OP and session, redirect the user to the
+            # beginning of login process without raising an error to the user. Attempt once.
+            logger.warning(f"No State Defined: {nsd_err}")
+            return redirect(request.session.get("next", "/"))
+        else:
+            # Clear the flag if the exception is not caught
+            request.session.pop("redirect_attempted", None)
+            return error_page(request, nsd_err)
```
Comment (abroddrick) [inline src/djangooidc/views.py:129]: **blocking question**

if I am reading this right, error_page(request, nsd_err) will call the error_page function but with error equal to the StateMismatch error, but in error_page it's checking if error is an instance of AuthenticationFailed or InternalError and then for all others it's throwing a 500. 

It looks like with this flow if a user ever gets to this line they will then see a 500, is that what we want?

I would think we want to 1.) just have StateMismatch inherit from AuthenticationFailed or 2.) also check for "is instance StateMatch) in error_page. 

Thoughts?

**Review by abroddrick — APPROVED** [review]
Just one blocking question, after that's answered I can approve

**Review by dave-kennedy-ecs — APPROVED** [review]
I have been unable to reproduce the original problem.  But I have reviewed the code and test code.  And I think safe to deploy.
I approve

File: src/djangooidc/views.py (hunk 8 above)
Comment (rachidatecs) [reply src/djangooidc/views.py:129]: Going with 1

**Pull request merged**